compiler/rustc_codegen_ssa/src/mir/place.rs

```diff
@@ -222,88 +224,121 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
+                    // this scalar (for either signed or unsigned), we can
+                    // identify whether this is a tagged variant with a single
+                    // comparison and no arithmetic.
+                    let (predicate, operand, constant) = if tags_start == low_unsigned {
+                        (IntPredicate::IntULE, niche, tags_end)
+                    } else if tags_end == high_unsigned {
+                        (IntPredicate::IntUGE, niche, tags_start)
+                    } else if tags_start == low_signed {
+                        (IntPredicate::IntSLE, niche, tags_end)
+                    } else if tags_end == high_signed {
+                        (IntPredicate::IntSGE, niche, tags_start)
                     } else {
-                        bx.intcast(relative_discr, cast_to, false)
+                        // That didn't work out, so we'll check
+                        // niche - tag_start <= relative-max
+                        // (requiring a subtraction and a comparison).
+                        let relative_discr =
+                            bx.sub(niche, bx.cx().const_uint_big(niche_llty, tags_start));
+                        (IntPredicate::IntULE, relative_discr, relative_max as u128)
                     };
-                    bx.add(
-                        relative_discr,
-                        bx.cx().const_uint(cast_to, niche_variants.start().as_u32() as u64),
-                    )
+
+                    let is_tagged =
+                        bx.icmp(predicate, operand, bx.cx().const_uint_big(niche_llty, constant));
+                    let cast_niche = bx.intcast(niche, cast_to, false);
+                    let delta = (tagged_variants.start().as_u32() as u128).wrapping_sub(tags_start);
+                    let tagged_discr = if delta == 0 {
+                        cast_niche
+                    } else {
+                        bx.add(cast_niche, bx.cx().const_uint_big(cast_to, delta))
+                    };
+                    (is_tagged, tagged_discr)
                 };
 
-                bx.select(
-                    is_niche,
-                    niche_discr,
+                let value = bx.select(
+                    is_tagged,
+                    tagged_discr,
                     bx.cx().const_uint(cast_to, untagged_variant.as_u32() as u64),
-                )
+                );
+
+                let lo = cmp::min(untagged_variant.as_u32(), tagged_variants.start().as_u32());
+                let hi = cmp::max(untagged_variant.as_u32(), tagged_variants.end().as_u32());
+                let lo = bx.cx().const_uint(cast_to, lo as u64);
+                let hi = bx.cx().const_uint(cast_to, hi as u64);
+                let lo_cmp = bx.icmp(IntPredicate::IntULE, lo, value);
+                let hi_cmp = bx.icmp(IntPredicate::IntULE, value, hi);
+                bx.assume(lo_cmp);
+                bx.assume(hi_cmp);
+                value
```

Comment:
Did you forget to apply most of this extra logic to cg_clif?

Reply:
I guess cf_clif refers to cranelift. Yeah I made no changes to code generation for cranelift.